drivers/stepper/adi_tmc/tmc2130_spi_step_dir.c

```diff
@@ -0,0 +1,374 @@
+	struct tmc2130_spi_sd_data *data = dev->data;
+	int ret;
+	uint32_t reg_value;
+
+	uint32_t ustep_res_reg_value = tmc2130_spi_sd_ms_res_translator(micro_step_res);
+
+	if (ustep_res_reg_value == 0xFF) {
+		return -EINVAL;
+	}
+
+	K_SPINLOCK((struct k_spinlock *)&data->common.lock) {
+		ret = tmc_spi_read_register(&config->spi, TMC2130_ADDRESS_MASK, TMC2130_CHOPCONF,
+					    &reg_value);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to read register 0x%x (error code: %d)", dev->name,
+				TMC2130_CHOPCONF, ret);
+			K_SPINLOCK_BREAK;
+		}
+		reg_value &= ~TMC2130_CHOPCONF_MRES_MASK;
+		reg_value |= (ustep_res_reg_value << TMC2130_CHOPCONF_MRES_SHIFT);
+
+		ret = tmc_spi_write_register(&config->spi, TMC2130_WRITE_BIT, TMC2130_CHOPCONF,
+					     reg_value);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to write register 0x%x (error code: %d)", dev->name,
+				TMC2130_CHOPCONF, ret);
+			K_SPINLOCK_BREAK;
+		}
+
+		data->ustep_res = micro_step_res;
+	}
+
+	return ret;
+}
+
+static int
+tmc2130_spi_sd_stepper_get_micro_step_res(const struct device *dev,
+					  enum stepper_micro_step_resolution *micro_step_res)
+{
+	struct tmc2130_spi_sd_data *data = dev->data;
+	*micro_step_res = data->ustep_res;
+	return 0;
+}
+
+static int tmc2130_spi_sd_stepper_move_to(const struct device *dev, int32_t target)
+{
+	struct tmc2130_spi_sd_data *data = dev->data;
+
+	if (!data->enabled) {
+		LOG_ERR("Failed to move to target position, device is not enabled");
+		return -ECANCELED;
+	}
+
+	return step_dir_stepper_common_move_to(dev, target);
+}
+
+static int tmc2130_spi_sd_stepper_move_by(const struct device *dev, int32_t steps)
+{
+	struct tmc2130_spi_sd_data *data = dev->data;
+
+	if (!data->enabled) {
+		LOG_ERR("Failed to move by delta, device is not enabled");
+		return -ECANCELED;
+	}
+
+	return step_dir_stepper_common_move_by(dev, steps);
+}
+
+static int tmc2130_spi_sd_stepper_run(const struct device *dev, enum stepper_direction direction)
+{
+	struct tmc2130_spi_sd_data *data = dev->data;
+
+	if (!data->enabled) {
+		LOG_ERR("Failed to run stepper, device is not enabled");
+		return -ECANCELED;
+	}
+
+	return step_dir_stepper_common_run(dev, direction);
+}
+
+static int tmc2130_spi_sd_stepper_init(const struct device *dev)
+{
+	const struct tmc2130_spi_sd_config *config = dev->config;
+	struct tmc2130_spi_sd_data *data = dev->data;
+	uint32_t gstat_data;
+	int ret;
+
+	uint8_t ustep_res_reg_value = tmc2130_spi_sd_ms_res_translator(data->ustep_res);
+	/* Dual Edge is configured in the same byte as microstep resolution */
+	if (config->common.dual_edge) {
+		ustep_res_reg_value += TMC2130_DOUBLE_EDGE_OFFSET;
+	}
+
+	/* Read GSTAT register to clear any errors. */
+	tmc_spi_read_register(&config->spi, TMC2130_ADDRESS_MASK, TMC2130_GSTAT, &gstat_data);
+	LOG_DBG("GSTAT: %x", gstat_data);
+
+	/* Check max values of properties. */
+	/* tpwmthrs is 20bit in length, so check the value. */
+	if (config->tpwmthrs > TPWMTHRS_MAX_VALUE) {
```

Comment:
I'd rather have those as `BUILD_ASSERTS` executed within the `DT_INST_FOR_EACH_COMPAT` so you have a build time check.